src/cmap/connection_pool.ts

```diff
@@ -544,7 +546,17 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
       fn(undefined, conn, (fnErr, result) => {
         if (typeof callback === 'function') {
           if (fnErr) {
-            callback(fnErr);
+            if ((fnErr as MongoError).code === MONGODB_ERROR_CODES.Reauthenticate) {
+              this.reauthenticate(conn, fn, (error, res) => {
+                if (error) {
+                  callback(error);
+                } else {
```

Comment:
It would be nice to remove the duplicated logic and the unnecessary nesting.  From slack - Durran is looking into a potential `withReauthentication` approach that can be used to wrap `fn` inside of `withConnection`.

Reply:
Added private `withReauthentication` now.